ls1intum/Artemis
# `Development`: Remove unnecessary fields from server response

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context


In #6318, @Strohgelaender noted that a method in `Course.java` (unrelated to the PR) wasn't annotated with `@JsonIgnore` although the client never needs the data produced by them. The server should not evaluate helper methods and send their results to the client as attributes that are never used.

### Description

I checked some methods in `Course`, `Exercise`, `Lecture`, `ProgrammingExercise`, `TextExercise` and `ModelingExercise` and found some more such instances. This PR adds the remaining `@JsonIgnore` annotations to avoid sending too many fields to the client.

### Steps for Testing
#### Test: Fields Removed
- Course field
  * Visit `artemisUrl/courses` and open DevTools.
  * Find the request to `/api/courses/for-dashboard` in the network tab.
  * Ensure that the field `validStartAndEndDate` does not exist on a course.
- Exercise fields
  * Open a new or existing programming exercise and your DevTools Network tab
  * Find the response to the request to `artemisUrl/api/exercises/<courseId>/details` and verify that these three fields are not included in the response: `released`, `branch`, `localSimulation`.
  * Open a new or existing modeling exercise and your DevTools Network tab
  * Find the response to the request to `artemisUrl/api/exercises/<courseId>/details` and verify that these two fields are not included in the response: `released`, `clusterBuildDate`.

#### Test: Everything Still Working
Because the fields that are `@JsonIgnore`d in this PR are never used in the client, the behavior should not change in any way. To test, attempt to find broken functionality around the following parts of the application:
- validity check of the start and end date of courses
- released state of exercises
- maybe some place where the branch name for programming exercises could be shown / where it could be shown whether the exercise is a local simulation
- maybe some place where the modeling cluster build date is used

#### For Code Tests
The code itself should be pretty straightforward to review. The difficult part is making sure that no client will break, most importantly the web client:
- For each of the fields that are `@JsonIgnore`d, try to find places in the client code that use it. AFAIK method names like `isValidStartAndEndDate` should be converted to the field name `validStartAndEndDate` (same for `get` as the prefix).
- Make an attempt to find out whether the fields might be in use for each of the following projects by ls1tum:
  * [Orion](https://github.com/ls1intum/Orion)
  * [Artemis iOS](https://github.com/ls1intum/artemis-ios)
  * [Artemis Android](https://github.com/ls1intum/artemis-android)
  * [Themis](https://github.com/ls1intum/Themis)
  * ~[Artemis SDK](https://github.com/ls1intum/artemis-sdk)~ @4ludwig4 told me that the code there is autogenerated and that it's currently no big deal to temporarily break it

#### Exam Mode Testing


Repeat the testing steps from above for programming exercises and modeling exercises within an exam.

### Review Progress






#### Code Review
#### Manual Tests
